// simple-react-form/lib/FieldType.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propTypes = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  /**
   * The current value of the field
   */
  value: _propTypes2.default.any,

  /**
   * Field label
   */
  label: _propTypes2.default.any,

  /**
   * The error message if there is a error
   */
  errorMessage: _propTypes2.default.string,

  /**
   * Call this function when the value changes
   */
  onChange: _propTypes2.default.func.isRequired,

  /**
   * If the input is disabled
   */
  disabled: _propTypes2.default.bool,

  /**
   * The schema for the field
   */
  fieldSchema: _propTypes2.default.object,

  /**
   * The schema for the object
   */
  schema: _propTypes2.default.object,

  /**
   * Use hint instead of label
   */
  useHint: _propTypes2.default.bool,

  /**
   * The name of the field
   */
  fieldName: _propTypes2.default.string.isRequired,

  /**
   * The props that must be passed to the child component
   */
  passProps: _propTypes2.default.object
};

var FieldType = function (_React$Component) {
  (0, _inherits3.default)(FieldType, _React$Component);

  function FieldType(props) {
    (0, _classCallCheck3.default)(this, FieldType);

    var _this = (0, _possibleConstructorReturn3.default)(this, (FieldType.__proto__ || Object.getPrototypeOf(FieldType)).call(this, props));

    console.warn("Now you don't have to extend FieldType to create components");
    return _this;
  }

  return FieldType;
}(_react2.default.Component);

exports.default = FieldType;


FieldType.propTypes = propTypes;
